plugins/redis/redis.cpp

```diff
@@ -0,0 +1,904 @@
+    assertOnError(pubReply->query(), "pub fail");
+    }
+#endif
+
+    //Check if returned value is locked
+    if (strncmp(reply->query()->str, REDIS_LOCK_PREFIX, strlen(REDIS_LOCK_PREFIX)) != 0)
+    {
+        //Not locked so return value
+        retVal->set(reply->query()->len, reply->query()->str);
+        return;
+    }
+    else
+    {
+        //Check that we SUBSCRIBEd to the correct channel (which could have been manually SET).
+        if (strcmp(reply->query()->str, channel) !=0 )
+        {
+            VStringBuffer msg("Redis Plugin: ERROR - the key '%s', on database %" I64F "u, is locked with a channel ('%s') different to that subscribed to.", key, database, reply->query()->str);
+            rtlFail(0, msg.str());
+            //MORE: We could attempt to recover at this stage by subscribing to the channel that the key was actually locked with.
+            //However, we may have missed the massage publication already or by then.
+            //If we ever changed the semantics of the 'timeout' to be that of these plugin functions rather than each redis call, we might as well
+            //subscribe again if there was time left on the clock.
+            //Since they are not, we could, though is this desirable behaviour?
+        }
+#if(0)//Added to allow for manual pub testing via redis-cli
+        struct timeval to = { 10, 0 };//10secs
+        redisSetTimeout(subConnection->context, to);
+#endif
+        //Locked so SUBSCRIBE
+        redisReply * nakedReply = NULL;
+        bool err = redisGetReply(subConnection->context, (void**)&nakedReply);
+        reply->setClear(nakedReply);
+        if (err != REDIS_OK)
+            rtlFail(0, "RedisPlugin: ERROR - GET timed out.");
+        assertOnCommandErrorWithKey(nakedReply, "GET", key);
+        if (nakedReply->type == REDIS_REPLY_ARRAY && strcmp("message", nakedReply->element[0]->str) == 0)
+        {
+            retVal->set(nakedReply->element[2]->len, nakedReply->element[2]->str);//return the published value rather than another (WATCHed) GET.
+            return;
+        }
+    }
+    throwUnexpected();
+}
+void Connection::handleLockOnSet(ICodeContext * ctx, const char * key, const char * value, size_t size, unsigned expire)
+{
+    StringBuffer cmd("SET %b %b");
+    RedisPlugin::appendExpire(cmd, expire);
+
+    //Due to locking logic surfacing into ECL, any locking.set (such as this is) assumes that they own the lock and therefore go ahead and set regardless.
+    //It is possible for a process/call to 'own' a lock and store this info in the LockObject, however, this prevents sharing between clients.
+    redisAppendCommand(context, cmd.str(), key, strlen(key), value, size);//SET
+    StringBuffer channel;
+    encodeChannel(channel, key);
+    redisAppendCommand(context, "PUBLISH %b %b", channel.str(), channel.length(), value, size);//PUB
```

Comment:
Can you assign to a variable, and then use in both commands or something similar using server side scripting?